### PR TITLE
Release planning for Mantid version 4.0

### PR DESCRIPTION
**Pull request by NickDraper**
A design document to capture the discussions around Mantid version 4.0

### CODE REVIEW

File: Design/Mantid4ReleasePlanning.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+User Expectations of Mantid 4.0
+-------------------------------
+
+We have not raised the concept of Mantid 4.0 with our users, so we have not made any commitment about what it may contain.  However we should be aware that while it is  not defined in Semantic versioning (http://semver.org/), the users will expect some significant improvements in order to justify the version change, and to act as a carrot to entice them to upgrade and overcome the discomfort of the breaking changes we are introducing.  Some of the breaking changes themselves will bring significant advantages, but the benefits of others are less visible.  It is worth considering other significant improvements that can be aligned with Mantid 4.0 to encourage enthusiastic adoption of these changes.
+
+List of breaking changes 
+------------------------
+
+1. Change from qwt5 graphs to ? [3 people 1-3m]
+1. Move to Python 3 on all platforms [1 person 2m]
```
Comment (quantumsteve) [inline Design/Mantid4ReleasePlanning.md]: Would you be more specific about what this change includes. Our current strategy is to maintain python 2.7 and python 3.x compatibility. Will there be a transition period where we ship separate packages for python 2 & 3? Can we realistically drop ALL support for python 2 with alongside the other substantial changes outlined here?

Comment (quantumsteve) [inline Design/Mantid4ReleasePlanning.md]: Building RHEL 7 RPMs for all Mantid dependencies using python 3 is a significant undertaking. @stuartcampbell might have a better idea of the time required.

File: Design/Mantid4ReleasePlanning.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+1. Sanitise Algorithm and property names (including merging) [Plan - 1 person 2w, Impl 3 people 1-3m]
+1. Removing InOut workspace usage [1 person 2w]
+1. ?Moving the Python API from the SimpleAPI (the impacts of this need further consideration) [????????]
```
Comment (SimonHeybrock) [inline Design/Mantid4ReleasePlanning.md]: - Remove old algorithm versions?

Comment (SimonHeybrock) [inline Design/Mantid4ReleasePlanning.md]: - Remove the old Python interface to workspaces, only provide the new one (removing, e.g., `readY()` and `dataY()`).

Comment (SimonHeybrock) [inline Design/Mantid4ReleasePlanning.md]: - It is likely that changes to indexing (https://github.com/mantidproject/documents/pull/24) will require removing things like `ISpectrum::setSpectrumNo()`, i.e., there is a series of breaking changes. This is a prerequisite for Instrument 2.0.

File: Design/Mantid4ReleasePlanning.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+List of other changes
+---------------------
+
+These changes are not bound to be included in this release, but being significant changes it would be good to include them together if
+the timing and resource allows.
+
+1. Instrument 2.0 [2p 4m]
```
Comment (SimonHeybrock) [inline Design/Mantid4ReleasePlanning.md]: - I think 2p 4m might be lower bound for roll out of the indexing changes I mention above, probably more, depending on the number of complications that arise.
- Currently I think it is likely that as a follow up and before we can introduce instrument 2.0 we might also need to cleanup workspace creation and initialization. Maybe 2p 2-4m.
- Instrument 2.0 itself can easily take 2p 6m for a basic implementation. This does *not* include advanced functionality, and does *not* include removing the old interface, i.e, refactoring existing algorithms comes on top of that (apart from simple cases).

Comment (OwenArnold) [inline Design/Mantid4ReleasePlanning.md]: I think 2p 6m for instrument 2.0 is a reasonable guess for `Instrument 2.0`. I would however much prefer that this estimate is kept private to the development team until the prototype is finished. I think I'll be able to give a much better estimate when we have approximations for all the features that we need built into the prototype.

**Comment by NickDraper** [comment]
Don't panic too much about the estimates at the moment for the PMB this afternoon I'll share a version with them with all of the estimates removed.

**Comment by NickDraper** [comment]
The PMB approved of the concept.  We need to flesh out what we plan to do, evaluate what should or should not be done for each area, the effort to implement, what approaches we can take to mitigate the change and the costs of these.

File: Design/Mantid4ReleasePlanning.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+User Expectations of Mantid 4.0
+-------------------------------
+
+We have not raised the concept of Mantid 4.0 with our users, so we have not made any commitment about what it may contain.  However we should be aware that while it is  not defined in Semantic versioning (http://semver.org/), the users will expect some significant improvements in order to justify the version change, and to act as a carrot to entice them to upgrade and overcome the discomfort of the breaking changes we are introducing.  Some of the breaking changes themselves will bring significant advantages, but the benefits of others are less visible.  It is worth considering other significant improvements that can be aligned with Mantid 4.0 to encourage enthusiastic adoption of these changes.
+
+List of breaking changes 
+------------------------
+
+1. Sanitise Algorithm and property names (including merging & generating naming convention) [2 person 8 months]
```
Comment (quantumsteve) [inline Design/Mantid4ReleasePlanning.md]: Exposing C++ getters/setters to python as [class properties](http://www.boost.org/doc/libs/1_63_0/libs/python/doc/html/tutorial/tutorial/exposing.html#tutorial.exposing.class_properties) would make the interface more pythonic. PR #18350 shows some of the opportunities & challenges.

**Comment by OwenArnold** [comment]
@NickDraper Why isn't this being merged. Don't we want to keep the outputs/decisions?

**Comment by NickDraper** [comment]
@OwenArnold Primarily this was killed by the user meeting of Jan 2017, or at least drastically changed when there was strong objection to almost all of the breaking changes.  There will not and has not been implemented in any particularly reasonable state. So it would seem strange to merge it as such.

**Comment by OwenArnold** [comment]
@NickDraper As you say, many of these changes were rejected in Jan 2017. Is there a separate record of that? I'm worried that this is the only document that would contain a record of what was originally proposed.

**Pull request closed without merging**